File: skyspark-scala/src/main/scala/com/akikanellis/skyspark/scala/api/algorithms/DominatingAlgorithm.scala
package com.akikanellis.skyspark.scala.api.algorithms

/**
  * An algorithm for determining if a point dominates another.
  */
private[algorithms] object DominatingAlgorithm {

  /**
    * Given two data points in a d-dimensional dataset, say p and q, we say p dominates q if for every i &isin; [1, d]
    * we have pi &le; qi and there exists at least one j such that pj &lt; qj .
    *
    * @param p The candidate point to check if it dominates the other point
    * @param q The other point to check if it is dominated by the candidate point
    * @return True if p dominates q, false if not
    */
  private[algorithms] def dominates(p: Point, q: Point): Boolean = {
    var atLeastOneDimensionIsSmaller = false

    for (i <- 0 until p.size) {
      val candidateDimension = p.dimension(i)
      val otherDimension = q.dimension(i)

      if (candidateDimensionIsWorse(candidateDimension, otherDimension)) return false

      if (candidateDimensionIsBetter(candidateDimension, otherDimension)) atLeastOneDimensionIsSmaller = true
    }

    atLeastOneDimensionIsSmaller
  }

  private def candidateDimensionIsWorse(candidateDimension: Double, otherDimension: Double): Boolean =
    candidateDimension > otherDimension

  private def candidateDimensionIsBetter(candidateDimension: Double, otherDimension: Double): Boolean =
    candidateDimension < otherDimension
}

File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/algorithms/DominatingAlgorithm.java
package com.akikanellis.skyspark.api.algorithms;

class DominatingAlgorithm {

    static boolean dominates(Point p, Point q) {
        boolean atLeastOneDimensionIsSmaller = false;

        for (int i = 0; i < p.size(); i++) {
            double candidateDimension = p.dimension(i);
            double otherDimension = q.dimension(i);

            if (candidateDimensionIsWorse(candidateDimension, otherDimension)) return false;

            if (candidateDimensionIsBetter(candidateDimension, otherDimension)) atLeastOneDimensionIsSmaller = true;
        }

        return atLeastOneDimensionIsSmaller;
    }

    private static boolean candidateDimensionIsWorse(double candidateDimension, double otherDimension) {
        return candidateDimension > otherDimension;
    }

    private static boolean candidateDimensionIsBetter(double candidateDimension, double otherDimension) {
        return candidateDimension < otherDimension;
    }
}
